### PR TITLE
Finetuning guide

### PR DESCRIPTION
**Pull request by strickvl**
Draft of documentation / guide around LLM finetuning.

Gitbook preview [here](https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/x9UKvIZt4T0daMw6mLMM/user-guide/llmops-guide/finetuning-llms/finetuning-100-loc).

## Remaining work

- [ ] more diagrams
- [ ] complete the deployment writeup (when #3032 is merged in)

### CODE REVIEW

**Review by htahir1 — CHANGES_REQUESTED** [review]
Alright i read it again. I think its great. But I have some feedback:

* There are lots of lists in the text like "Here is a list: here is an explanation" -> Id format this with the first part being in bold "**Here is a list**: here is an explanation"
* This is again a pet peeve but bullet points , should they begin with a capital or small letter?...
* For me the chapter that sort of doesn't "fit" well into the guide is "Starter choices with fine-tuning". I'm not quite sure whether its the lack of link outs or the lack of code in there that makes it just feel like I have a lot of reading to do before I get to the action.. maybe there is a way to condense it into a visual element or maybe tell the story in a code-centric way? Or give more examples as to the choices available? 
* Finetuning with Accelerate doesn't point out that we have an integration with a decorator... that is what I expected here tbh
* I expected a link from the new evaluation section to the old RAG evaluation section but I found only "Please note that the approach taken for these evaluations is similar to that used and [showcased in the RAG guide](https://github.com/o/-MCl1Hlw9oU4xibZ-ymf/s/x9UKvIZt4T0daMw6mLMM/user-guide/llmops-guide/evaluation)." .. Maybe there is a way to pick up the story from there? What is actually different from what was said before? There might be a shortcut by making a link back
* The deployment chapter is missing a code snippet from our VLLM deployer by @dudeperf3ct :-) Id love to get that in there, even if it isn't released yet.

File: docs/book/user-guide/llmops-guide/finetuning-llms/finetuning-llms.md
Comment (htahir1) [inline docs/book/user-guide/llmops-guide/finetuning-llms/finetuning-llms.md:1]: This seems underdeveloped. Also dont we call this README.md usually?

File: docs/book/user-guide/llmops-guide/finetuning-llms/why-and-when-to-finetune-llms.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    R -->|No| L
+    Q -->|No| S
+```
+
+## Alternatives to consider
+
+Before deciding to finetune an LLM, consider these alternatives:
+
+- Prompt engineering: Often, carefully crafted prompts can achieve good results without the need for finetuning.
+- [Retrieval-Augmented Generation (RAG)](../rag-with-zenml/understanding-rag.md): For many use cases involving specific knowledge bases, RAG can be more effective and easier to maintain than finetuning.
```
Comment (htahir1) [inline docs/book/user-guide/llmops-guide/finetuning-llms/why-and-when-to-finetune-llms.md:77]: broken link

File: docs/book/user-guide/llmops-guide/finetuning-llms/getting-started-with-finetuning.md
Comment (htahir1) [inline docs/book/user-guide/llmops-guide/finetuning-llms/getting-started-with-finetuning.md:1]: the name of the file can be a bit more aligned with the content

File: docs/book/user-guide/llmops-guide/finetuning-llms/getting-started-with-finetuning.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+The data needed for your use case will follow directly from the specific use case you're choosing, but ideally it should be something that is already *mostly* in the direction of what you need. It will take time to annotate and manually transform data if it is too distinct from the specific use case you want to use, so try to minimise this as much as you possibly can.
+
+A couple of examples of where you might be able to reuse pre-existing data:
+
+- you might have examples of customer support email responses for some specific scenario which deal with a well-defined technical topic that happens often but that requires these custom responses instead of just a pro-forma reply
+- you might have manually extracted metadata from customer data or from business data and you have hundreds or (ideally) thousands of examples of these
+
+In terms of data volume, a good rule of thumb is that for a result that will be rewarding to work on, you probably want somewhere in the order of hundreds to thousands of examples.
+
+### 🎯 Good vs Not-So-Good Use Cases
```
Comment (htahir1) [inline docs/book/user-guide/llmops-guide/finetuning-llms/getting-started-with-finetuning.md:58]: i love this, it deserves a blog post on its own

File: docs/book/user-guide/llmops-guide/finetuning-llms/next-steps.md
Comment (htahir1) [inline docs/book/user-guide/llmops-guide/finetuning-llms/next-steps.md:1]: IMO we can link to papers / books or other resources that might be useful here

**Review by htahir1 — CHANGES_REQUESTED** [review]
Feedback:

- Looks like a lot of files have been deleted - would it make sense to add redirects for these?
- Other than that and the comments this looks SUPER amazing and solid. Great work!!!

**Comment by strickvl** [comment]
@htahir1 the deleted files were actually duplicate docs which weren't linked anywhere, so no need to link. they weren't in our TOC to start with.

File: docs/book/user-guide/llmops-guide/finetuning-llms/getting-started-with-finetuning.md
Comment (strickvl) [reply docs/book/user-guide/llmops-guide/finetuning-llms/getting-started-with-finetuning.md:1]: done

**Review by htahir1 — APPROVED** [review]
Amazing! LGTM!

File: docs/book/user-guide/llmops-guide/finetuning-llms/why-and-when-to-finetune-llms.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+- LLM finetuning is not a universal solution or approach: it won't and cannot solve every problem, it might not reach the required levels of accuracy or performance for your use case and you should know that by going the route of finetuning you are taking on a not-inconsiderable amount of technical debt.
+- Chatbot-style interfaces are not the only way LLMs can be used: there are lots of uses for LLMs and this finetuning approach which don't include any kind of chatbot. What's more, these non-chatbot interfaces should often to be considered preferable since the surface area of failure is much lower.
+- The choice to finetune an LLM should probably be the final step in a series of experiments. As with the first point, you shouldn't just jump to it because other people are doing it. Rather, you should probably rule out other approaches (smaller models for more decomposed tasks, [RAG](../rag-with-zenml/understanding-rag.md) if you're working on a retrieval or long-context problem, or a mixture of the above for more complete use cases).
+
+## When makes sense to finetune an LLM?
```
Comment (wjayesh) [inline docs/book/user-guide/llmops-guide/finetuning-llms/why-and-when-to-finetune-llms.md:14]:
```suggestion
## When does it make sense to finetune an LLM?
```

**Review by wjayesh — APPROVED** [review]
